controller/api/destination/server.go

```diff
@@ -383,11 +427,38 @@ func getSvcID(k8sAPI *k8s.API, clusterIP string, log *logging.Entry) (*watcher.S
 	return service, nil
 }
 
-// getPod returns a pod that maps to the given IP address. The pod can either
+// getPodByHostname returns a pod that maps to the given hostname (or an
+// instanceID). The hostname is generally the prefix of the pod's DNS name;
+// since it may be arbitrary we need to look at the corresponding service's
+// Endpoints object to see whether the hostname matches a pod.
+func getPodByHostname(k8sAPI *k8s.API, hostname string, svcID watcher.ServiceID) (*corev1.Pod, error) {
+	ep, err := k8sAPI.Endpoint().Lister().Endpoints(svcID.Namespace).Get(svcID.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, subset := range ep.Subsets {
+		for _, addr := range subset.Addresses {
+			if addr.TargetRef == nil || addr.TargetRef.Kind != "Pod" {
+				continue
+			}
+
+			if hostname == addr.Hostname {
+				podName := addr.TargetRef.Name
+				podNamespace := addr.TargetRef.Namespace
+				return k8sAPI.Pod().Lister().Pods(podNamespace).Get(podName)
```

Comment:
The pod IP indexer from `InitializeIndexers` should have indexed this pod by it's IP. I think we should be able to use `addr.ip` here and get it from that indexer so that we can avoid using a lister.

Reply:
Similar to my comment above, we know the pod name and namespace.  Why would we use an index when we can retrieve it directly?

Reply:
Yeah, both Kevin and I changed our opinion on this after an off GitHub conversation (see my reply above). It makes more sense to retrieve the pod directly here through the Lister. 👍🏻